### PR TITLE
Solution chapter2

### PR DESCRIPTION
**Pull request by capital-G**
### Solutions for Chapter {NUMBER}

cc @vrom911 @chshersh

Still problems w/ building happy under OSX (see https://github.com/capital-G/learn4haskell/pull/1)

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,8 +348,10 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
-
+subList a b x
+  | a<0 || b <0 = []
```
Comment (vrom911) [inline src/Chapter2.hs]: That's right!
In Haskell functions and operators are always separated from the arguments with spaces 👾 

```suggestion
  | a < 0 || b < 0 = []
```

File: src/Chapter2.hs [hunk 2]
```diff
@@ -336,8 +348,10 @@ from it!
...
+  | otherwise = take (b-a+1) (drop a x)
```
Comment (vrom911) [inline src/Chapter2.hs]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check above :slightly_smiling_face:

File: src/Chapter2.hs [hunk 3]
```diff
@@ -606,7 +622,8 @@ Implement a function that duplicates each element of the list
 
 -}
 duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
+duplicate [] = []
+duplicate (x:xs) = [x,x] ++ duplicate xs
```
Comment (vrom911) [inline src/Chapter2.hs]: It is more efficient to construct lists using the cons operator directly without introducing intermediate values 🐌 🙂 
```suggestion
duplicate (x:xs) = x : x : duplicate xs
```

File: src/Chapter2.hs [hunk 4]
```diff
@@ -621,7 +638,10 @@ Write a function that takes elements of a list only in even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = [] 
+takeEven (x:_:xs) = [x] ++ takeEven xs
```
Comment (vrom911) [inline src/Chapter2.hs]: Wonderful, great pattern-matching skills 💯 

It's better to use `:` here as well:

```suggestion
takeEven (x:_:xs) = x : takeEven xs
```

File: src/Chapter2.hs [hunk 5]
```diff
@@ -728,7 +748,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = concat (map (\x -> replicate x x) l)
```
Comment (vrom911) [inline src/Chapter2.hs]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 6]
```diff
@@ -781,13 +801,14 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
 
--- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Int -> [Int] -> [Int]
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: [Int] -> [Int] -> [Int]
+pairMul = zipWith (*)
```
Comment (vrom911) [inline src/Chapter2.hs:811]: Eta-reduction award :trophy:

File: src/Chapter2.hs [hunk 7]
```diff
@@ -842,7 +863,10 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate n xs
+  | n < 0 = []
+  | otherwise = take (length xs) $ drop n $ cycle xs
```
Comment (vrom911) [inline src/Chapter2.hs]: That's a correct solution! 

Though, it does a little bit more than work when the given number is big and the list is small. You will cycle through the list a lot. There is a way to optimize this using `mod` 😉  :slightly_smiling_face:

File: src/Chapter2.hs [hunk 8]
```diff
@@ -858,7 +882,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = rewind xs ++ [x]
```
Comment (vrom911) [inline src/Chapter2.hs:888]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by vrom911 — APPROVED** [review]
Great work on the Chapter 🥳 
Very thoughtful solutions 👌🏼

File: src/Chapter2.hs (hunk 8 above)
Comment (capital-G) [reply src/Chapter2.hs:888]: Thanks for the hint - I will keep this in mind but for now I think I'll keep it at the simple implementation :) One step at a time

**Pull request merged**